profiles/common/modules/features/nexteuropa_newsroom/classes/NewsroomAgenda.inc

```diff
@@ -0,0 +1,679 @@
+
+    if (is_object($topics)) {
+      $this->setTopicUrl($topics->name);
+      $topics = array($topics);
+    }
+
+    $this->setTopics($topics);
+
+    $this->addUrlPart($this->getView());
+    $this->addUrlPart($this->getTypeUrl());
+    $this->addUrlPart($this->getTopicUrl());
+  }
+
+  /**
+   * Prepare agenda components: query, result, the dates range.
+   */
+  protected function prepareAgenda() {
+    $this->getBaseQuery();
+    $this->setResult($this->prepareResult($this->loadResult()));
+    $this->setupAgenda();
+  }
+
+  /**
+   * Returns previous title.
+   *
+   * @return string
+   *   Title
+   */
+  public function getPreviousTitle() {
+    return t('Previous');
+  }
+
+  /**
+   * Returns next title.
+   *
+   * @return string
+   *   Title
+   */
+  public function getNextTitle() {
+    return t('Next');
+  }
+
+  /**
+   * Returns form id with the current ID.
+   *
+   * @param string $form_id
+   *   Form ID.
+   *
+   * @return array
+   *   Form array.
+   */
+  protected function getForm($form_id) {
+    $topics = $this->getTopics();
+    $topic_id = count($topics) == 1 ? $topics[0]->tid : NULL;
+    $result_topics = $form_id == 'nexteuropa_newsroom_agenda_filter_form' ? $this->getResultTopicsId() : NULL;
+    return drupal_get_form($form_id, $this->getView(), $this->getTypeId(), $topic_id, $this->getCurrentDate()->format('Y-m-d'), $result_topics);
+  }
+
+  /**
+   * Returns the agenda html.
+   *
+   * @return string
+   *   HTML.
+   */
+  public function getAgenda() {
+    $cache_key = 'get_agenda:page:' . $this->getUniqueCachePart();
+    $cache = cache_get($cache_key, NEXTEUROPA_NEWSROOM_CACHE_TABLE);
+    if ($cache) {
+      $content = $cache->data;
+    }
+    else {
+      $this->prepareAgenda();
+      $content = theme('newsroom_agenda_page', array(
+        'filter_form' => $this->getForm('nexteuropa_newsroom_agenda_filter_form'),
+        'date_form' => $this->getDisplayCalendar() ? $this->getForm('nexteuropa_newsroom_agenda_datepicker_form') : NULL,
+        'items' => $this->getCurrentItems(),
+        'next_event_items' => $this->getNextItems(),
+        'is_block' => $this->getIsBlock(),
+        'navigation' => array(
+          'previous' => $this->getPreviousDate() ? $this->getNavigationLink($this->getPreviousTitle(), $this->getPreviousDate()) : NULL,
+          'next' => $this->getNextDate() ? $this->getNavigationLink($this->getNextTitle(), $this->getNextDate()) : NULL,
+        ),
+      ));
+      cache_set($cache_key, $content, NEXTEUROPA_NEWSROOM_CACHE_TABLE, time() + 3600);
+    }
+
+    return $content;
+  }
+
+  /**
+   * Generate content for Agenda block.
+   *
+   * @return string
+   *   Content of the block.
+   */
+  public function getBlockAgenda() {
+    $cache_key = 'get_agenda:block:' . $this->getUniqueCachePart();
+    $cache = cache_get($cache_key, NEXTEUROPA_NEWSROOM_CACHE_TABLE);
+    if ($cache) {
+      $content = $cache->data;
+    }
+    else {
+      $this->prepareAgenda();
```

Comment:
I have updated the code according to your first comment 
`
/**
   * Returns the agenda html.
   *
   * @return string
   *   HTML.
   */
  public function getAgenda() {
    $this->prepareAgenda();
    $filter_form = $this->getForm('nexteuropa_newsroom_agenda_filter_form');
    $date_form = $this->getForm('nexteuropa_newsroom_agenda_datepicker_form');
    $date_form_output = $this->getDisplayCalendar() ? drupal_render($date_form) : NULL;
    $content = theme('newsroom_agenda_page', array(
      'filter_form' => drupal_render($filter_form),
      'date_form' => $date_form_output,
      'items' => $this->prepareItemsOutput($this->getCurrentItems()),
      'next_event_items' => $this->prepareItemsOutput($this->getUpcomingItems()),
      'is_block' => $this->getIsBlock(),
      'navigation' => array(
        'previous' => $this->getPreviousDate() ? $this->getNavigationLink($this->getPreviousTitle(), $this->getPreviousDate()) : NULL,
        'next' => $this->getUpcomingDate() ? $this->getNavigationLink($this->getNextTitle(), $this->getUpcomingDate()) : NULL,
      ),
    ));
    return $content;
  }
`
Forms are not cached anymore, and **prepareAgenda** returns the cached data.

What else is wrong here ?